posts/lint-rcpp/index.qmd

```diff
@@ -0,0 +1,168 @@
+
+_cpplint_ can be run locally from the terminal, and instructed to not lint `RcppExports.cpp`, but all other `.cpp` files in `src/`.
+
+```sh
+cpplint --exclude="src/RcppExports.cpp" src/*.cpp
+```
+
+### Cppcheck
+
+[_cppcheck_](https://cppcheck.sourceforge.io/) is a static code analysis tool, that aims to "have very few false positives". This is especially useful for the non-standard organisation of Rcpp projects compared to C++ projects.
+
+_cppcheck_ can also be run locally and instructed to ignore the autogenerated `RcppExports.cpp` file, while throwing up issues with style.
+
+```sh
+cppcheck -i src/RcppExports.cpp --enable=style --error-exitcode=1 src
+```
+
+Here, the `--enable=style` option lets _cppcheck_ flag issues with style, acting as a second linter.
+Enabling all checks (`--enable=all`) would flag issues with the `Rcpp.h` header not being found (of the class `missingIncludeSystem`), or with the solver functions not being used (`unusedFunction`). 
+
+These extra options should be avoided in Rcpp packages, as the linking is handled for us, and the functions are indeed used later --- just not by other C++ functions.
+
+## Adding C++ linting to CI workflows
+
+Both _cpplint_ and _cppcheck_ can be easily added to continuous integration workflows. In the Epiverse, we use Github Actions. The C++ lint workflow we have implemented looks like this:
+
+```yaml
+on:
+  push:
+    paths: "src/**"
+  pull_request:
+    branches:
+      - "*"
+
+name: Cpp-lint-check
+
+jobs:
+  cpplint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v2
+      - run: pip install cpplint
+      - run: cpplint --quiet --exclude="src/RcppExports.cpp" src/*.cpp
+
+  cppcheck:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - run: sudo apt-get install cppcheck
+      - run: cppcheck -isrc/RcppExports.cpp --quiet --enable=warning --error-exitcode=1 .
+```
+
+The workflow is triggered when there are changes to files in `src/`, and on all pull requests.
+
+## Formatting C++ code
+
+C++ code can be automatically formatted to avoid linter errors.
+An especially useful tool is [_clang-format_](https://clang.llvm.org/docs/ClangFormat.html). Our code is styled to follow the [Google C++ style guide](https://google.github.io/styleguide/cppguide.html) using:
+
+```sh
+clang-format -i -style=google src/*.cpp
+```
+
+However, this also formats the autogenerated `RcppExports.cpp` file.
+It can be extra work to repeatedly undo this change and keep the original formatting, or to repeatedly re-style the file as it's rewritten after each call to `Rcpp::compileAttributes()`.
+One solution is to include stylers in a pre-commit action called a ['git hook'](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks), so that the file is always styled before committing, and differences do not reflect formatting changes.
```

Comment:
I don't think this is a good approach.

It is generally disproved to style / reformat auto-generated files (https://github.com/r-lib/styler/issues/215). This creates unnecessary and confusing diffs for external contributors. For example, I was confused why I kept getting diffs with style changes on `RcppExports.cpp` in finalsize.

Since we already exclude `RcppExports.cpp` from the checks, why not exclude it from the reformatting either?

Your issue may be that `clang-format` does not provide an argument to exclude a specific file but you can use the following command:

```sh
find src -name "*.cpp" ! -name "RcppExports.cpp" -exec clang-format -stle=google -i {} \;
```

Reply:
Yes, I stopped formatting `RcppExports.cpp` some time ago after initially formatting it. I can add this as a way to exclude RcppExports from formatting. There's another way (supposedly) using `.clang-format-ignore`, but I haven't been able to get it to work.

Reply:
Fixed in https://github.com/epiverse-trace/epiverse-trace.github.io/pull/7/commits/c2da24228d857320605ca609753a13f71441aaa3